apps/emqx_rule_engine/include/rule_engine.hrl

```diff
@@ -61,28 +61,49 @@
         }).
 
 -record(action_instance,
-        { id :: action_instance_id()
-        , name :: action_name()
-        , fallbacks :: list(#action_instance{})
-        , args :: #{binary() => term()} %% the args got from API for initializing action_instance
+        { id
```

Comment:
Curious: why should we separate record definition from its type?

Reply:
Because of matchspec usages, when we construct a record with specs like `'_'`, dialyzer complains that the values don't match the typespec. For example:

```
****: Record construction #rule{id::'_',for::'_',rawsql::'_',is_foreach::'_',fields::'_',doeach::'_',incase::'_',conditions::'_',on_action_failed::'_',actions::'_',enabled::'_',created_at::'_',description::'_',state::'_'} violates the declared type of field id::binary() and for::[binary()] and rawsql::binary() and is_foreach::boolean() and fields::[any()] and incase::[any()] and conditions::tuple() and on_action_failed::'continue' | 'stop' and actions::[#action_instance{id::binary(),name::atom(),fallbacks::[#action_instance{id::binary(),name::atom(),fallbacks::[any()],args::#{binary()=>_}}],args::#{binary()=>_}}] and enabled::boolean() and created_at::integer() and description::binary()
```